client/app/bundles/course/reference-timelines/components/TimeBar/TimeBar.tsx

```diff
@@ -0,0 +1,151 @@
+import { useEffect, useState } from 'react';
+import moment from 'moment';
+
+import { DAY_WIDTH_PIXELS, getDaysFromWidth } from '../../utils';
+import HorizontallyDraggable from '../HorizontallyDraggable';
+
+import DurationBar from './DurationBar';
+
+interface TimeBarProps {
+  startsAt: string | moment.Moment;
+  bonusEndsAt?: string | moment.Moment;
+  endsAt?: string | moment.Moment;
+  shadow?: boolean;
+  disabled?: boolean;
+  selected?: boolean;
+  onClick?: (target: HTMLElement | null) => void;
+  onChangeTime?: (newTime: TimeTriplet, rollback: () => void) => void;
+}
+
+interface TimeTriplet {
+  start: moment.Moment;
+  bonus?: moment.Moment;
+  end?: moment.Moment;
+}
+
+const isValidTimeTriplet = ({ start, bonus, end }: TimeTriplet): boolean => {
+  const isBonusSameOrAfterStart = bonus?.isSameOrAfter(start);
+  const isEndSameOrAfterBonus = bonus ? end?.isSameOrAfter(bonus) : undefined;
+  const isEndAfterStart = end?.isAfter(start);
+
+  return (
+    (isBonusSameOrAfterStart ?? true) &&
+    (isEndSameOrAfterBonus ?? true) &&
+    (isEndAfterStart ?? true)
+  );
+};
+
+const isSameTimeTriplet = (time1: TimeTriplet, time2: TimeTriplet): boolean => {
+  const isStartSame = time1.start.isSame(time2.start);
+
+  const isBonusSame =
+    time1.bonus === time2.bonus || Boolean(time1.bonus?.isSame(time2.bonus));
+
+  const isEndSame =
+    time1.end === time2.end || Boolean(time1.end?.isSame(time2.end));
+
+  return isStartSame && isBonusSame && isEndSame;
+};
+
+const generateTriplet = (
+  start: TimeBarProps['startsAt'],
+  bonus: TimeBarProps['bonusEndsAt'],
+  end: TimeBarProps['endsAt'],
+): TimeTriplet => ({
+  start: moment(start),
+  bonus: bonus ? moment(bonus) : undefined,
+  end: end ? moment(end) : undefined,
+});
+
+const TimeBar = (props: TimeBarProps): JSX.Element => {
+  const [{ start, bonus, end }, setTime] = useState<TimeTriplet>(
+    generateTriplet(props.startsAt, props.bonusEndsAt, props.endsAt),
+  );
+
+  const [dragging, setDragging] = useState(false);
+  const [oldTime, setOldTime] = useState<TimeTriplet>();
+
+  useEffect(() => {
```

Comment:
Because `TimeTriplet` is stored as a state in the component. I use `useEffect` to listen to changes to time from `props`, e.g., when the time was updated from the store.

Reply:
Got it! I am trying to think a way to try to not use `useState` + `useEffect` (ie use `useMemo`) but seems like we need it since we need `validateAndSetTime` and `handleChangeTime`